docs/architecture/adr-011-reports-module.md

```diff
@@ -0,0 +1,302 @@
+# ADR 011: Reports module
+
+## Changelog
+
+- April 12th, 2022: Initial draft
+
+## Status
+
+DRAFT
+
+## Abstract
+
+This ADR contains the specification of the `x/reports` module which will allow users to report other users and/or content posted inside subspaces.
+
+## Context
+
+One of the most important parts of a functional social network is the set of rules that constitute the Terms of Service, and how they are enforced. In most social networks the key for a good user experience is the reporting system: users are able to report misbehaving users or inappropriate content that gets later reviewed and taken care of by moderators. 
+
+Inside Desmos we MUST give users a set of tools that allow them to perform the same actions on all the social networks that they will have access to. At the same time, we SHOULD NOT enforce any high-level Term of Service since each subspace that will be built on Desmos might want to have very different rules from the other ones (i.e. an adult content social network will most likely have a very different set of rules from a kids social network). Instead, we should allow subspace owners and admins to register the various reasons a user/content can be reported for.
+
+## Decision
+
+We will implement a new module named `x/reports` that allows users to report either a misbehaving user or a bad content inside subspaces where they have the permission to do so.
+
+The same module will also allow subspace owners to register their own custom supported reasons for reports. 
+
+### Types
+Reports can be of two different types: `UserReport` for reporting a misbehaving user, or `PostReport` for reporting a bad post. The whole system must be thought in order to support additional report types in the future (i.e. `ProposalReport` to report a bad on-chain governance proposal). 
+
+#### Report
+```protobuf
+// Report contains the data of a generic report
+message Report {
+  // Id of the report inside the subspace
+  required uint64 id = 1;
+
+  // Target of the report
+  oneof Target {
+    UserData user_data = 2;
+    PostData post_data = 3;
+  }
+  
+  // Id of the reason this report has been created for
+  required uint32 reason_id = 4;
+  
+  // Message attached to this report
+  optional string message = 5;
+  
+  // Address of the reporter 
+  required string reporter = 6;
+}
+
+// UserData contains the data of a report about a user
+message UserData {
```

Comment:
What's the reason behind using a wrapper for the `user` and the `post_id`?
I can only thing about extending the data contained here purposes but honestly I don't know
what else can be added here.

Reply:
It's both for extensibility and readability. 

Another way of doing this would have been to have: 
- an enum that contains all the possible values 
- a generic `data` field that can be used to specify things

However, the `data` would have needed to be something like `[]byte` to support integers (post ids) and strings (user addresses). Then we would have needed to read those in different ways based on the type of the report. 

Honestly I find it a lot more easy to manage and read having wrapping classes: it makes it very easy to integrate without requiring much documentation and special cases.

Reply:
You're right, they didn't make sense. I have now reviewed the `UserData` and `PostData` to also contain the id of the subspace inside which the user should be reported or the reported post is

Reply:
@RiccardoM Great! This made me think about another kind of report that we might want to consider, the `SubspaceReport`. Does it make sense to you? I'm struggling cause from one side, it could be useful to let people and clients know which are the those dangerous subspaces (like those promoting hatread, violence, racism) and from another one it could lead to what in the cinema and video games industries is called "review-bombing" (in our case **"reports-bombing"**) having users reporting a subspace just because it doesn't match their ideas or beliefs.

Reply:
@bragaz Honestly I don't think it's worth implementing it. We should allow for max freedom when creating subspaces, just like how the internet is. Also, we won't have any way of acting on top of subspaces anyway: we can't delete them so even if users report them what can we do?

Reply:
Nothing, true.